Documentation on custom package repositories

jonasfj opened the pull request:
Authentication with tokens is new in 2.15, which sort of makes how to use custom package repositories a bit more interesting.

This was just my attempt at writing something that helps people understand custom package repositories, without having to piece together the different bits from all over. 

Staged: https://kw-staging-dartlang-2.web.app/tools/pub/custom-package-repositories

parlough commented on src/tools/pub/dependencies.md:
```diff
@@ -114,8 +114,8 @@ This example specifies that your package depends on a hosted package named
 
-If you want to use your own package server, you can use a description that
-specifies its URL:
+If you want to use your [own package server](./custom-package-repository),
```
```suggestion
If you want to use your [own package server](/custom-package-repositories),
```

parlough commented on src/tools/pub/custom-package-repositories.md:
```diff
@@ -0,0 +1,119 @@
+_hosted-url_ (e.g. `https://pub.example.com/my-account/`). 
+
+Third-party package servers can be useful for hosting private packages. It is
+also common to use [git-dependencies](./dependencies#git-packages) for hosting
```
```suggestion
also common to use [git-dependencies](/dependencies#git-packages) for hosting
```

jonasfj replied:
I'm guessing that if we want to make links absolute we have start from `/tools/...`

parlough commented on src/tools/pub/custom-package-repositories.md:
```diff
@@ -0,0 +1,119 @@
+private package repository.
+
+
+## Authenticating with a custom package repository
+Custom package servers are often private and require authentication. The `pub`
```
```suggestion
## Authenticating with a custom package repository

Custom package servers are often private and require authentication. The `pub`
```

parlough commented on src/tools/pub/custom-package-repositories.md:
```diff
@@ -0,0 +1,119 @@
+use the `dart pub token add <hosted-url>` command.
+
+
+## Getting dependencies from a custom package repository
+To fetch a package from custom package repository we must specify the
```
```suggestion
## Getting dependencies from a custom package repository

To fetch a package from custom package repository we must specify the
```

parlough commented on src/tools/pub/custom-package-repositories.md:
```diff
@@ -0,0 +1,119 @@
+_hosted-url_ for the package in `pubspec.yaml`, using the syntax for
+[hosted packages](./dependencies#hosted-packages).
+
+{% prettify yaml tag=pre+code %}
+dependencies:
+  foo:
+    hosted: https://pub.example.com
+    version: ^1.4.0
+{% endprettify %}
```
````suggestion
```yaml
dependencies:
  foo:
    hosted: https://pub.example.com
    version: ^1.4.0
```
````

jonasfj replied:
I didn't know we could do that, we should probably patch the other pages -- I just cargo culted.. :D

parlough commented on src/tools/pub/custom-package-repositories.md:
```diff
@@ -0,0 +1,119 @@
+token.
+
+
+### Using multiple package repositories
+It is possible to fetch different dependencies from different package
```
```suggestion
### Using multiple package repositories

It is possible to fetch different dependencies from different package
```

parlough commented on src/tools/pub/custom-package-repositories.md:
```diff
@@ -0,0 +1,119 @@
+repositories, as the _hosted-url_ is specified for each dependency. As
+illustrated below.
+
+{% prettify yaml tag=pre+code %}
+dependencies:
+  # package retry is fetched from pub.dev (the default package repository)
+  retry: ^3.0.0
+  # package foo is fetched from https://pub.example.com
+  foo:
+    hosted: https://pub.example.com
+    version: ^1.4.0
+{% endprettify %}
```
````suggestion
```yaml
dependencies:
  # package retry is fetched from pub.dev (the default package repository)
  retry: ^3.0.0
  # package foo is fetched from https://pub.example.com
  foo:
    hosted: https://pub.example.com
    version: ^1.4.0
```
````

parlough commented on src/tools/pub/custom-package-repositories.md:
```diff
@@ -0,0 +1,119 @@
+and either update the `dependencies` section in section in each package, or
+override the default package repository.
+
+{{site.alert.note}}
+To ensure that public packages are usable to everyone, the official package
+repository [pub.dev](https://pub.dev) does not allow publication of packages
+with git-dependencies or hosted-dependencies with custom package repositories.
+
+However, such packages can be published to a custom package repository!
+{{site.alert.end}}
```
```suggestion
{{site.alert.note}}
  To ensure that public packages are usable to everyone, the official package
  repository [pub.dev]({{site.pub}}) does not allow publication of packages
  with git-dependencies or hosted-dependencies with custom package repositories.

  However, such packages can be published to a custom package repository!
{{site.alert.end}}
```

parlough commented on src/tools/pub/custom-package-repositories.md:
```diff
@@ -0,0 +1,119 @@
+{{site.alert.end}}
+
+
+### Overriding the default package repository
+The package repository to be used can be specified on a per-dependency basis,
```
```suggestion
### Overriding the default package repository

The package repository to be used can be specified on a per-dependency basis,
```

parlough commented on src/tools/pub/custom-package-repositories.md:
```diff
@@ -0,0 +1,119 @@
+### Overriding the default package repository
+The package repository to be used can be specified on a per-dependency basis,
+or the default package repository can be overriden with the
+[`PUB_HOSTED_URL`](./environment-variables) environment variable.
```
```suggestion
[`PUB_HOSTED_URL`](/environment-variables) environment variable.
```

parlough commented on src/tools/pub/custom-package-repositories.md:
```diff
@@ -0,0 +1,119 @@
+
+This can be useful if you are mirroring all packages in a private package
+repository, or working in a restricted network environment using a mirror of
+[pub.dev](https://pub.dev).
```
```suggestion
[pub.dev]({{site.pub}}).
```

parlough commented on src/tools/pub/custom-package-repositories.md:
```diff
@@ -0,0 +1,119 @@
+[pub.dev](https://pub.dev).
+
+
+## Publishing to a custom package repository
+The `pub` tool also supports publishing packages to a private package
```
```suggestion
## Publishing to a custom package repository

The `pub` tool also supports publishing packages to a private package
```

parlough commented on src/tools/pub/custom-package-repositories.md:
```diff
@@ -0,0 +1,119 @@
+
+## Publishing to a custom package repository
+The `pub` tool also supports publishing packages to a private package
+repository, using the [`publish_to`](./pubspec#publish_to) property in
```
```suggestion
repository, using the [`publish_to`](/pubspec#publish_to) property in
```

parlough commented on src/tools/pub/custom-package-repositories.md:
```diff
@@ -0,0 +1,119 @@
+repository, using the [`publish_to`](./pubspec#publish_to) property in
+`pubspec.yaml`. If working on a private package it is a good idea to specify
+this early in the development, so as to prevent accidental publication to
+[pub.dev](https://pub.dev). 
```
```suggestion
[pub.dev]({{site.pub}}). 
```

parlough commented on src/tools/pub/custom-package-repositories.md:
```diff
@@ -0,0 +1,119 @@
+To publish a package to `https://pub.example.com` you would write a
+`pubspec.yaml` as follows, and run `dart pub publish`.
+
+{% prettify yaml tag=pre+code %}
+name: mypkg
+version: 1.0.0
+# Ensures the package is published to https://pub.example.com
+publish_to: https://pub.example.com
+{% endprettify %}
```
````suggestion
```
name: mypkg
version: 1.0.0
# Ensures the package is published to https://pub.example.com
publish_to: https://pub.example.com
```
````

parlough commented on src/tools/pub/custom-package-repositories.md:
```diff
@@ -0,0 +1,119 @@
+publish_to: https://pub.example.com
+{% endprettify %}
+
+{{site.alert.note}}
+Even if you don't have a private repository, you can specify `publish_to: none`
+which stops accidental publication.
+{{site.alert.end}}
```
```suggestion
{{site.alert.note}}
  Even if you don't have a private repository, 
  you can specify `publish_to: none`
  which stops accidental publication.
{{site.alert.end}}
```

parlough commented on src/tools/pub/custom-package-repositories.md:
```diff
@@ -0,0 +1,119 @@
+{{site.alert.end}}
+
+
+## Writing a custom package repository
+The REST API for writing a custom package repository is outlined in the
```
```suggestion
## Writing a custom package repository

The REST API for writing a custom package repository is outlined in the
```

parlough commented on src/tools/pub/custom-package-repositories.md:
```diff
@@ -0,0 +1,119 @@
+## Getting dependencies from a custom package repository
+To fetch a package from custom package repository we must specify the
+_hosted-url_ for the package in `pubspec.yaml`, using the syntax for
+[hosted packages](./dependencies#hosted-packages).
```
```suggestion
[hosted packages](/dependencies#hosted-packages).
```

parlough reviewed (commented):
Some formatting and link suggestions.

kwalrath commented on src/tools/pub/custom-package-repositories.md:
```diff
@@ -0,0 +1,119 @@
+---
+title: Custom Package Repositories
```
```suggestion
title: Custom package repositories
```

kwalrath commented:
Thanks for the page, @jonasfj! We'll do a copy edit pass once you've addressed other reviewers' feedback.

jonasfj commented:
Thanks, I think I addressed all comments. Feel free to edit as much as you please :D

I don't know if an entry in the sidebar is warranted, I didn't add it as I wasn't sure it was necessary (or where to put it).
